internal/safebuffer/safereader/safereader.go

```diff
@@ -0,0 +1,160 @@
+package safereader
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// Reader is a goroutine safe io.ReadCloser that makes initial data available to
+// callers as well as new data that arrives later via channel.
+type Reader struct {
+	mu        sync.Mutex
+	buf       *bytes.Buffer
+	close     func()
+	closed    <-chan struct{}
+	dataCh    <-chan []byte
+	dataAvail chan struct{}
+	closeOnce sync.Once
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// jobIsRunning doesn't have a race because Buffer.NewReader calls it while
+// holding a lock that prevents new data, or the close of jobDone, from being
+// processed.
+func jobIsRunning(jobDone <-chan struct{}) bool {
+	select {
+	case <-jobDone:
+		return false
+	default:
+		return true
+	}
+}
+
+type Channels struct {
+	Data   chan []byte
+	Closed <-chan struct{}
+}
+
+// New returns a new Reader that keeps a copy of data and will append any data
+// that arrives on the returned Data channel. If the job is no longer running
+// when created, it is treated as a closed reader and no new data can be added.
+// Calls to Close() in this case are ignored.
+func New(data []byte, jobDone <-chan struct{}) (*Reader, *Channels) {
+	closed := make(chan struct{})
+	closeFn := func() { close(closed) }
+
+	buf := make([]byte, len(data))
+	copy(buf, data)
+
+	r := Reader{
+		close:  closeFn,
+		closed: closed,
+		buf:    bytes.NewBuffer(buf),
```

Comment:
Why do we need to keep a full copy of the job output per reader?

If a job outputs 1Gb of data and we have 5 active readers, we will take 5GB of ram. Is there a way where we do not need to implement a solution with such bottleneck ?

Reply:
+1

Reply:
I tried to call this out in the pull request's description, but I should also explain it in the code. This is by far the biggest compromise I've made in the design. Keeping a full copy of the output buffer from the time that a reader is created is a very naive approach that is very memory and time inefficient. I did this to keep the code simple and with the understanding the the challenge was given with "unlimited cpu and unlimited memory". This is not a design I would ever propose for production.

I think what I would have liked to do instead is use a `RWMutex` on the original buffer and let the readers read directly from that and maintain their own position/index. The problem I had when thinking through this is that if I imagined a lot of connected readers holding `RLock`s, then writes may block for unacceptably long times. I may be able to fix this by limiting the time that `RLock`s are held and looping, thereby giving an opportunity for the write locks to be acquired. I worried that trying to figure out how to do this correctly and without races would become quite complex. In keeping with the spirit of the challenge, "do the simplest thing possible that satisfies the requirements", I chose what I thought to be the simpler option to implement.

There is another optimization I could also do to the current implementation, which would be to discard data from readers once it has been read. Again, it's more work and not necessarily a requirement, so I did not choose to implement this either.